advanced-programming/mutexes/my_mutex.go

```diff
@@ -0,0 +1,62 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type mutex struct {
+	val int64
+}
+
+var counter int
+
+func (m *mutex) Lock() {
+	for {
+		if atomic.CompareAndSwapInt64(&m.val, 0, 1) == true {
+			return
+		}
+	}
+}
+
+func (m *mutex) Unlock() {
+	swapped := atomic.CompareAndSwapInt64(&m.val, 1, 0)
+	if swapped == false {
+		fmt.Println("tried to unlocked an unlocked mutex")
+		return
+	}
+
+	return
+}
+
+func acqAndRelease(i int, m *mutex, wg *sync.WaitGroup) {
+	// Acquire mutex
+	m.Lock()
+	fmt.Printf("Goroutine %d: The mutex at address %p is acquired\n", i, &m)
+
+	// do work
+	time.Sleep(1 * time.Second)
```

Comment:
Aside: In a real system, you'd likely want the lock to be held for as short a period as possible. In an ideal situation you'd have something like this:

```
m.Lock()
// shared queue
work := workQueue.take()
m.Unlock()

// do slow work WITHOUT mutex held
time.Sleep(1 * time.Second)

...
```

Reply:
@robot-dreams Thanks for this suggestion. Makes total sense. Enjoyed this exercise.